### DISCUSSION TITLE
redesign the tvm model

### DISCUSSION THREAD

**Discussion by Alec4r** [eduNEXT/tvm#8] [Ideas]
# How is working now?
![current_tvm](https://user-images.githubusercontent.com/40800297/168336375-941d2f52-15fa-46a1-a14b-a7ee298623d8.png)

Currently, we need a step sequence.

## To use tvm in local project
**In a directory (example/stack):**
1. virtualenv venv
2. Install tvm
3. `tvm setup` this will create a symlink from **example/stack/.tvm/switcher** to **venv/bin/tutor**
4. Deactivate and active virtualenv
5. `tvm install v12.2.0`, this will clone the tag in example/stack/.tvm/v12.2.0
6. `tvm use v12.2.0`, this will update the switcher file (example/stack/.tvm/switcher)
7. `tutor config printroot`, this should return the tutor config default root (~.local/share/tutor)

You only can run tutor and tvm commands if you have enabled the virtualenv.

## To use tvm global
**In a directory (example/stack):**
1. virtualenv venv
1. Install tvm
2. `tvm setup -g` this will create a symlink from **example/stack/.tvm/switcher to** **usr/local/bin/tutor**
3. `tvm install v12.2.0`, this will clone the tag in example/stack/.tvm/v12.2.0
4. `tvm use v12.2.0`, this will update the switcher file (example/stack/.tvm/switcher)
5. `tutor config printroot`, this should return the tutor config default root (~.local/share/tutor)

You can run tutor and tvm commands from anywhere.

# What is the problem?
With this infrastructure, we have different problems:

1. To use tvm in a project is necessary to set up tvm in the project directory, this means that you need to create one .tvm directory in each project directory and clone there each version again.

2. If you want to set up it global and for a reason, you removed the project you will break all tutor commands, the reason why this happens is that you didn't remove the symlink.

3. You will always need to install tvm in a virtualenv to set up local.

4. To set up the tutor root, you need to export manually the TUTOR_ROOT variable.

# What is the proposal?
![newtvmproposal](https://user-images.githubusercontent.com/40800297/168392341-f6b27af3-4947-464a-aa50-fd4b4e38093d.png)

## Remove setup -g command
The plan is that all projects and versions will be cloned in a unique folder, I think that is a good idea to clone them in **~/.tvm**

So in the first command or in the install process, we will create the files that need to tvm work.

## Using tvm project init and tvm project enable instead of tvm setup
This command will create a .tvmrc file with the settings on the project directory, something like `git init`

The file will have next settings:
- tutor-root (current path)
- tutor-plugins (current path + /plugins)
- version (tvm project version)

```bash
tvm project init edunext --version="v12.2.0"
```

If you don't define the flag version in the command, tvm will create a new tvm project version with the name <tvm_version_in_use>@<project_name>

```
tvm project init tests
e.g. v12.2.0@tests
```

Then you should call the command `tvm project enable` this command will active the virtualenv created in the tvm project and it will set TUTOR_ROOT and TUTOR_PLUGIN_ROOT

## Running pip commands
In both cases, you should call `tvm packages [options]` to install pip packages

```bash
tvm packages install tutor-mfe
```

# What will be the new step by step?

## Global
1. Install tvm `pip install git+https://github.com/eduNEXT/tvm.git`
2. `tvm install v12.2.0`, this will clone the tag in **~/.tvm/v12.2.0**
3. `tvm use v12.2.0`, this will update the switcher file (**~/.tvm/switcher**)
4. `tutor config printroot`, this should return the tutor config default root (~.local/share/tutor)

## Local
1. Install tvm `pip install git+https://github.com/eduNEXT/tvm.git`
2. `tvm install v12.2.0`, this will clone the tag in **~/.tvm/v12.2.0**
3. `tvm use v12.2.0`, this will update the switcher file (**~/.tvm/switcher**)
4. In your project directory `tvm init example`, this will create a .tvmrc file 
   with version v12.2.0@example.
5. `tvm enable`, this will activate the tvm project's virtualenv.
6. `tutor config printroot`, this should return your project path.

**Comment by Alec4r** [comment]
# Video (ES)
https://user-images.githubusercontent.com/40800297/168806793-3227e207-2592-43e8-8625-6c28c618d4b6.mp4

**Comment by felipemontoya** [comment]
This is great @Alec4r.

I like quite a lot the mechanism to have one single `~/.tvm` location where all versions are installed. I also love the @project_name solution to manage the plugins_root.

The fact that you have a working prototype is also amazing.

I would like to make this possible by making smaller scope PRs that solve one issue at the time.

For instance:
- a PR that makes the global install at ~/.tvm possible
- a PR that makes creates the new usage of version@project_name. I would even make a default name for project so that not writing a name does not result in an error.
- a PR for switching the json file to the .tvmrc file.
- some more...

**Comment by Alec4r** [comment]
This has been merged #10
